src/backend/joanie/payment/models.py

```diff
@@ -1 +1,465 @@
+from django.utils.translation import get_language
+from django.utils.translation import gettext_lazy as _
+
+from babel.numbers import get_currency_symbol
+from djmoney.models.fields import MoneyField
+from djmoney.money import Money
+from howard.issuers import InvoiceDocument
+from parler.utils import get_language_settings
+from parler.utils.context import switch_language
+
+from joanie.core.models import Order
+from joanie.core.utils import merge_dict
+
+from . import enums as payment_enums
+from . import get_payment_backend
+
+User = get_user_model()
+
+
+class Invoice(models.Model):
+    """
+    Invoice model is an accounting element related to an order
+    """
+
+    parent = models.ForeignKey(
+        "self",
+        on_delete=models.RESTRICT,
+        related_name="children",
+        verbose_name="parent",
+        blank=True,
+        null=True,
+    )
+    reference = models.CharField(
+        db_index=True,
+        editable=False,
+        max_length=40,
+        unique=True,
+    )
+    created_on = models.DateTimeField(
+        _("created on"),
+        auto_now_add=True,
+        editable=False,
+    )
+    updated_on = models.DateTimeField(
+        _("updated on"),
+        auto_now=True,
+        editable=False,
+    )
+    order = models.OneToOneField(
+        Order,
+        verbose_name=_("order"),
+        on_delete=models.PROTECT,
+        blank=True,
+        null=True,
+    )
+    total = MoneyField(
+        _("total"),
+        max_digits=9,
+        decimal_places=2,
+        default_currency=settings.DEFAULT_CURRENCY,
+        null=False,
+        blank=False,
+    )
+    recipient_name = models.CharField(
+        _("invoice recipient"), max_length=40, null=False, blank=False
+    )
+    recipient_address = models.TextField(
+        _("invoice address"), max_length=255, null=False, blank=False
+    )
+    localized_context = models.JSONField(
+        _("context"),
+        help_text=_("Localized data that needs to be frozen on invoice creation"),
+        editable=False,
+    )
+
+    class Meta:
+        db_table = "joanie_invoice"
+        verbose_name = _("Invoice")
+        verbose_name_plural = _("Invoices")
+        constraints = [
+            models.CheckConstraint(
+                check=(
+                    models.Q(order__isnull=False)
+                    & models.Q(parent__isnull=True)
+                    & models.Q(total__gte=0)
+                )
+                | (models.Q(order__isnull=True) & models.Q(parent__isnull=False)),
+                name="invoice_relies_on_order_or_invoice",
+            ),
+        ]
+
+    def __str__(self):
+        types = dict(payment_enums.INVOICE_TYPES)
+        return f"{types[self.type]} {self.reference}"
+
+    @property
+    def state(self):
+        """
+        Process the state of the invoice
+        """
+        if self.balance.amount >= 0:
+            if (
+                self.invoiced_balance.amount == 0
```

Comment:
Should it be D('0.00') everywhere? Make sure it's well tested.

Reply:
It has been tested.
https://github.com/openfun/joanie/blob/3d1ff89c424f6de92c25965efc02b2aba93f32f9/src/backend/joanie/payment/tests/test_models_invoice.py#L137-L184

We can compare a Decimal to an Integer and I found the syntax clearer. But I agree we loose the fact that amount in a Decimal.

Reply:
It's ok if D("0.00") == 0 but I wasn't sure.